bin/i18n/translation_status/update_translation_status.rb

```diff
@@ -4,20 +4,103 @@
+  string_keys = redshift_client.exec(unique_string_key_query).map {|row| row['string_key']}
+
+  # Used to determine if a given string_key is translated or not.
+  translation_service = TranslationService.new
+  # All the locale codes we support i.e. en_US, es_MX, de_DE, etc.
+  locales = Languages.get_locale.map {|lang| lang[:locale_s]}
+  # Record what time we checked the translation status for these strings. Use the same time for all the entries
+  # so we can tell which batch they were all checked in.
+  # Timestamp format matches the one supported by Redshift.
+  current_time = Time.now.utc.strftime("%Y-%m-%d %H:%M:%S")
+  # Loop through each locale code our platform supports, i.e. en_US, es_MX, de_DE, etc. We want to check the translation
+  # status in each language for every string_key.
+  locales.each_with_index do |locale, i|
+    puts "Analyzing translation status of #{locale} (#{i + 1}/#{locales.size})"
+    translation_statuses = []
+    # Record the translation status for every string_key, i.e. "data.script.name.express-2020.title"
+    string_keys.each do |string_key|
+      # Maybe there is bad data in the database and we recorded an empty string
+      next if string_key.blank?
+      translated = translation_service.translated?(locale, string_key)
+      translation_statuses << {
+        string_key: string_key,
+        locale: locale,
+        has_translation: translated,
+        checked_at: current_time
+      }
+    end
+    # Create the rows to be inserted in the analysis.i18n_string_translation_status Redshift table
+    # https://docs.aws.amazon.com/redshift/latest/dg/c_Examples_of_INSERT_30.html
+    values = translation_statuses.map do |status|
+      # Escape special SQL characters to protect against SQL injection attacks.
+      ActiveRecord::Base.sanitize_sql(
+        ["('%s', '%s', '%s', '%s')",
+         status[:string_key],
+         status[:locale],
+         status[:has_translation],
+         status[:checked_at]]
+      )
+    end.join(',')
+    # Run a Redshift query to insert the translation statuses into the analysis.i18n_string_translation_status table.
+    query = <<~SQL.squish
+      INSERT INTO analysis.i18n_string_translation_status (string_key, locale, has_translation, checked_at)
+      VALUES #{values}
+    SQL
+    redshift_client.exec(query)
+  end
+end
+
+# @param [Integer] day_count The number of days in the past to look for unique string_key's
+# @return [String] A SQL query which returns the unique string_key's logged in the past day_count days.
+def unique_string_key_sql_query(day_count)
+  <<~SQL.squish
     SELECT DISTINCT string_key
     FROM analysis.i18n_string_tracking_events
     WHERE environment = 'production'
     AND created_at >= current_timestamp - interval '#{day_count} days'
   SQL
-
-  redshift_client.exec(query).map {|row| row['string_key']}
 end
+
+# This script is scheduled to run regularly on the i18n-dev server.
+update_translation_status(90)
```

Comment:
Could you add explanation for the value 90? It isn't clear to me why we have to go back 3 months if this script runs everyday.

Reply:
Oh I was doing this because the existing data is pretty old. I turned off the logging because we have some pending performance improvements assigned to me this sprint. I will make sure to check in a more reasonable time like 7 days or something.